GVFS/GVFS.Platform.Windows/WindowsFileSystem.cs

```diff
@@ -32,6 +35,63 @@ public void ChangeMode(string path, int mode)
         public bool TryGetNormalizedPath(string path, out string normalizedPath, out string errorMessage)
         {
             return WindowsFileSystem.TryGetNormalizedPathImplementation(path, out normalizedPath, out errorMessage);
+        }
+
+        public bool TryReadFirstByteOfFile(string fileName)
+        {
+            return NativeFileReader.TryReadFirstByteOfFile(fileName);
+        }
+
+        private class NativeFileReader
+        {
+            private const uint GenericRead = 0x80000000;
+            private const uint OpenExisting = 3;
+
+            public static bool TryReadFirstByteOfFile(string fileName)
+            {
+                using (SafeFileHandle handle = Open(fileName))
+                {
+                    if (!handle.IsInvalid)
+                    {
+                        return ReadOneByte(handle);
+                    }
+                }
+
+                return false;
+            }
+
+            private static SafeFileHandle Open(string fileName)
+            {
+                return CreateFile(fileName, GenericRead, (uint)(FileShare.ReadWrite | FileShare.Delete), 0, OpenExisting, 0, 0);
+            }
+
+            private static unsafe bool ReadOneByte(SafeFileHandle handle)
+            {
+                byte[] buffer = new byte[1];
+                int bytesRead = 0;
+                fixed (byte* p = buffer)
+                {
+                    return ReadFile(handle, p, 1, bytesRead, 0);
+                }
+            }
+
+            [DllImport("kernel32", SetLastError = true, ThrowOnUnmappableChar = true, CharSet = CharSet.Unicode)]
+            private static extern SafeFileHandle CreateFile(
+                string fileName,
+                uint desiredAccess,
+                uint shareMode,
+                uint securityAttributes,
+                uint creationDisposition,
+                uint flagsAndAttributes,
+                int hemplateFile);
+
+            [DllImport("kernel32", SetLastError = true)]
+            private static extern unsafe bool ReadFile(
+                SafeFileHandle file,
+                [Out] byte* buffer,
+                int numberOfBytesToRead,
+                [Out] int numberOfBytesRead,

```

Comment:
How did this translate to an int? Shouldn't it be a pointer?

Reply:
I think what's happening is that the type is not translated correctly here, but because the value you're passing in happens to be initialized to 0, the native code is probably seeing this as a null pointer and ignoring it. But anything other than a 0 will probably cause a crash.

Reply:
And I can see now that this is the pre-existing Windows code, and I may have been the person who wrote this incorrect code :-)